csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Assets.cs

```diff
@@ -0,0 +1,162 @@
+        ///   UsedPackages = {
+        ///     "castle.core",
+        ///     "json.net"
+        ///   }
+        /// </summary>
+        private Dependencies AddPackageDependencies(JObject json, Dependencies dependencies)
+        {
+            // If there are more than one framework we need to pick just one.
+            // To ensure stability we pick one based on the lexicographic order of
+            // the framework names.
+            var references = json
+                .GetProperty("targets")?
+                .Properties()?
+                .MaxBy(p => p.Name)?
+                .Value
+                .ToObject<Dictionary<string, ReferenceInfo>>();
+
+            if (references is null)
+            {
+                progressMonitor.LogDebug("No references found in the targets section in the assets file.");
+                return dependencies;
+            }
+
+            // Find all the compile dependencies for each reference and
+            // create the relative path to the dependency.
+            return references
+                .Aggregate(dependencies, (deps, r) =>
+                {
+                    var info = r.Value;
+                    var name = r.Key.ToLowerInvariant();
+                    if (info.Type != "package")
+                    {
+                        return deps;
+                    }
+
+                    // If this is a .NET framework reference then include everything.
+                    return netFrameworks.Any(framework => name.StartsWith(framework))
+                        ? deps.Add(name, "")
+                        : info
+                            .Compile
+                            .Aggregate(deps, (d, p) => d.Add(name, p.Key));
+                });
+        }
+
+        /// <summary>
+        /// Parse `json` as project.assets.json content and add relative paths to the dependencies
+        /// (together with used package information) required for compilation.
+        /// </summary>
+        /// <returns>True if parsing succeeds, otherwise false.</returns>
+        public bool TryParse(string json, Dependencies dependencies)
+        {
+            try
+            {
+                var obj = JObject.Parse(json);
+                AddPackageDependencies(obj, dependencies);
+                return true;
+            }
+            catch (Exception e)
+            {
+                progressMonitor.LogDebug($"Failed to parse assets file (unexpected error): {e.Message}");
+                return false;
+            }
+        }
+
+        public static Dependencies GetCompilationDependencies(ProgressMonitor progressMonitor, IEnumerable<string> assets)
+        {
+            var parser = new Assets(progressMonitor);
+            return assets.Aggregate(new Dependencies(), (dependencies, asset) =>
+            {
+                var json = File.ReadAllText(asset);
+                parser.TryParse(json, dependencies);
+                return dependencies;
+            });
```

Comment:
Personal preference: I would avoid `Aggregate` and use a `foreach` instead. `Dependencies` is basically a collection (`List<>`), and we rarely see the pattern:
```C#
return assets.Aggregate(new List<int>(), (list, asset) => {});
```

Reply:
In this particular case, I think it makes sense to use Aggregate, but to streamline it with the other part of the implementation I will change it to `ForEach`.